### ISSUE TITLE
implementation of stale bot

### ISSUE THREAD

**Issue by lub** [Mailu/Mailu#1642]
Hi all,

as some of you already noticed we implemented a stale bot for issues (PRs are unaffected!) in this repository.

I can assure you, that the reason for this is not to close all issues or otherwise just empty the issue tracker.

We are currently way overdue for 1.8, way behind reviewing PRs and also don't have any overview about which issues are actually important and which could simply be closed.

Of course the logical reaction is "just review all issues!11", and we are currently doing this, as some of you (hopefully) also noticed. But going forward with the current maintainer bandwidth we don't see it as realistic to regularily review all open issues for e.g. stale bug reports or unclear feature requests and these things are currently littering the open issues.

There are some labels exempted from that stale handling. The reasoning is basically that every issue that seems fixable should get one of these labels:
https://github.com/Mailu/Mailu/blob/master/.github/stale.yml#L14

If you get a message from stale bot on one of your issues and you think it was a mistake, just leave a comment on it. As the number of actually stale and closable issues probably decreases in the coming days/weeks such mistakenly stale issues should hopefully catch our attention much faster anyway.
It's also possible possible for us to re-open an issue, in case stale bot or a human closed it and you think it was a mistake.

Also feel free to come to our Matrix room (https://matrix.to/#/#mailu:tedomum.net) and discuss such issues there, as it can of course happen, that we for example dimiss the notification/open an issue and then get distracted by something else in our lifes. After all this project is run by volunteers in their free time, so it's sadly not always possible to give everything the attention it probably deserves.

**Comment by Nebukadneza** [comment]
Let’s continue the discussion from #1451 in here …

> IMO it's not right to close an issue just because currently no one can work on it. It does not become a non-issue because of that and if you close it some one will come and open a new one. There is no benefit is closing issues without fixing just for the sake of closing.

I’m inclined to agree, however, i think there is a little more to consider: We have to keep a overview of what’s currently important, happening and "to do" as humans. As the mountain of issues keeps growing, it’s very easy to lose sight of the overall picture — a situation which leaves the maintainers paralyzed and unable to act. The idea of this voting-round and the stale bot is to keep only things visible that fulfill 2 points:
* They are needed so much that they are being actively requested and/or discussed
* They are actively being worked on
*We do realize that issues not falling in the above category can be valid as well.* However, there needs to be a way to sort things into categories. The 2 most important categories we currently think of are:
* Things that we can imagine to review, integrate or implement with the current "workforce"
* Things we imagine would be too big to handle, in either reviewing/integrating or implementation
This is the most harsh way of sorting. The "softer" label-mechanism have proven not to be enough to manage the mountain of possible work in the past. Hence, we resort to tools like the stale-bot, which do help in some larger projects.

To summarize: It is not that a stale/closed issue is one we don’t view as valid. It’s one we cannot imagine possible to integrate with the time we have at hands currently.

One more thing to note: Any user or contributor serious about an issue would search for both open and closed issues. Should someone find a closed-due-to-stale issue they really want to implement, re-opening is the correct mechanism, and I don’t view this as something bad.

> it's not right to close an issue just because currently no one can work on it. It does not become a non-issue because of that and if you close it some one will come and open a new one. There is no benefit is closing issues without fixing just for the sake of closing

As explained above, making it possible for a human contributor to judge an issues state is the benefit we hope to get from it. It’s neither a non-issue because of that, nor is it "for the sake of it": We have a specific goal in mind with this, and that is to make the whole thing tractable again.

> So if what you're trying to do is somehow make Mailu **_look complete_** because there are no open issues, this will be effective. If you want Mailu to be the best mail server out there, don't be so anxious to close issues that people spent good time thinking through and articulating. It's rude and people will start to look elsewhere for solutions.

We are not trying to make anything look better than reality. We are trying to make the state of our heap-of-work map the reality: Even sorting issues and reviewing pull-requests has shown to be more work than the current team can handle. That is the sad reality, and we need to take some kind of action since we need to map to the state of issues/PRs to the reality.

I hope this can explain why we’re doing things to you.

**Comment by Nebukadneza** [comment]
> And I see that `/lifecycle frozen` has been disabled.

Nothing has been disabled. These commands seem to be a custom bot common in the k8s repositories, and one of their custom services. It’s neither common nor advisable to assume all github projects use the same set of bots and integrations, just because one organization uses them …

**Comment by briantopping** [comment]
Stalebot is for closing bugs that have probably self-resolved, not features that are not implemented.

Ideas for good features will remain ideas fo good features as long as they are not closed and forgotten. 

Ideas for bad features should be closed by someone with karma. Not a bot because some arbitrary length of time has passed.

Either fix / enable `/lifecycle frozen`, get a different bot or stop using them.

This current mindset is wrong and is toxic. It will disenfranchise the community from taking action.

There's another discussion going in parallel in #1451. Feedback welcome in either.

**Comment by Nebukadneza** [comment]
Please move all your discussion here. Having the same conversation in multiple places is not a good idea, and will derail the respective issues into off-topic.



> Stalebot is for closing bugs that have probably self-resolved, not features that are not implemented.

I think you fundamentally misunderstood what we want to do with this: Our goal is not to have stale-bot remove all issues, especially feature-requests, that linger around. Currently, if you look throught the recently labelled or closed issues, we are going through all of them manually, assign label to whatever is a legitimate request that we feel is in-line with the project goals. These have labels that prevent them from being closed — consult the stalebot config for details.

> Ideas for good features will remain ideas fo good features as long as they are not closed and forgotten.

We are in agreement on this point, and the mechanism of doing so is to assign one of the labels that prevent stalebot from hitting. Issues with:
* a priority
* marked as backlog
* marked as actively being worked on
Will not be targeted by the bot.

> Either fix / enable /lifecycle frozen, get a different bot or stop using them.

Again, this is something very specific to one bot in one project. It’s not a generally accepted mechanism or tool. If you want to feel at home and use the same tools everywhere, please supply details. If you want to have a open discussion about "what is right": I’m open and very interested in that.

My gut-feeling is: That tools frozen-command prevents any kind of auto-close action as soon as anyone with permission invokes it. This is exactly the same machanism that we implement using the abovementioned labels. If you think there is a better mechanism, please explain what it should do, and why it’s better.


After reading all the users and contributors comments, I do realize one thing: We did not communicate the true goal of this tool properly. That is a mistake we made, and I am sorry for that. Please let me remedy this:

### Why do we have a stale-bot? What is going to become of all those issues?
In the past, the Mailu-Project had phases in which the maintainers and contributors were so busy, that sorting incoming issues was not possible. This lead to a situation where no productive work was possible for a long time. We want to avoid getting in these situations.

To reach that goal, we have worked on a couple of things:
* Add issue templates so that only well-described bugs or feature-requests get tickets. This reduces the flood of incoming issues. User-support are welcome in the matrix chat.
* Use milestones for releases and reduce the label-jungle: So that getting an overview is easier, and it’s clearer to contributors and maintainers how to tag issues and PRs
* Implement a stale-bot to work in 2 situations: Closing actually-stale bugs or questions that are never going to get resolved ; Being a *emergency* stopper in the hopefull-never-again-phases where noone can even *look at* incoming issues.
* Manually triage open issues in a 2-3 people meeting, to subsequently label issues we think are legitimate properly, so that they are not going to get auto-closed.

We have that with these measures in place, this toxic extreme-overload situation will cease to happen. If it should happen, we at least have a tool that informs issue-authors there is something wrong, and at the same time helps (albeit in a slightly ugly fashion) keep only the most important stuff afloat. The latter point is important: When after such a phase contributors and maintainers return, they are not paralyzed and driven-away by an unmanageable mountain of work.



I hope the above explanation made our goals clearer. Again: I am sorry for our lack of communication, we could have done better.

**Comment by AndrewSav** [comment]
@lub, I'm reposting [from here](https://github.com/Mailu/Mailu/issues/783#issuecomment-699180818) as per @Nebukadneza request. Please let me know your thoughts.

> IMO it's similarly rude to expect members of the project to regularly go through old issues that could've been long closed by the original authors. 

Sorry, I have no idea what you mean here. If you do not want to regularly go through the issues than do not, but closing a legitimate issue, just because you cannot work on them right now is still not nice toward people who contributed to them regardless of whether you  regularly go through the issues or not, the point still stands.

>I'm not saying that this is something you did, but when you look through the issues there are many that couldn't be reproduced and are lacking feedback from the very same authors.

This is easy to solve, ask for clarification, if not received, close as cannot reproduce. But keep closing stuff because there were "no activity" will just lead to endless +1's and `/make unstale`. It's not like an issue needs a constant drip of information, once it's clear what it's about.

**Comment by AndrewSav** [comment]
@Nebukadneza 

> Implement a stale-bot to work in 2 situations: Closing actually-stale bugs or questions that are never going to get resolved ; Being a emergency stopper in the hopefull-never-again-phases where noone can even look at incoming issues.

Could you please clarify both these situations in more details?

> Closing actually-stale bugs or questions that are never going to get resolved

Currently the stale bot simply closes after some inactivity time. Inactivity does not indicate that the issue is "actually-stale bugs or questions that are never going to get resolved". It could indicated a few things. For example it could indicated that the dev team did not have time to triage or prioritise it. In this latter case, do you think it's a right thing to close the issue, and if yes, why?

> Being a emergency stopper in the hopefull-never-again-phases where noone can even look at incoming issues.

How does this work? Could you please provide some more details?

**Comment by AndrewSav** [comment]
@Nebukadneza 

I appreciate your write up, but I feel that certain things need to be spelt out - a general write up while useful to understand general direction / bigger picture does not address those details.

Brian wrote:

> Stalebot is for closing bugs that have probably self-resolved, not features that are not implemented.
> Ideas for good features will remain ideas fo good features as long as they are not closed and forgotten.
> Ideas for bad features should be closed by someone with karma. Not a bot because some arbitrary length of time has passed.
> Either fix / enable /lifecycle frozen, get a different bot or stop using them.

Could you please comment on the above, in particular:

* Is stale bot going to continue to close issue because some arbitrary length of time has passed?
* What are going to be the rules / setting for the stale bot?
* Will you allow your user base to `/lifecycle frozen`, when you yourselves do not have time to look / triage? 

It sounds like the team does not have a lot of time, so I'm worried that issues might get closed before the team have a change to look at them.

> My gut-feeling is: That tools frozen-command prevents any kind of auto-close action as soon as anyone with permission invokes it. This is exactly the same mechanism that we implement using the abovementioned labels.

The main difference is that if  `/lifecycle frozen` worked _I_ could trigger it, and the abovementioned labels _you_ have to trigger. So I feel I'm left without any agency at all here. To re-iterate, I feel it is okay, to close an issue (after a discussion perhaps), the dev team believes won't fit. It is not okay to close it, because some  arbitrary length of time has passed. Would you agree with that?

**Comment by Nebukadneza** [comment]
> > Implement a stale-bot to work in 2 situations: Closing actually-stale bugs or questions that are never going to get resolved ; Being a emergency stopper in the hopefull-never-again-phases where noone can even look at incoming issues.
> 
> Could you please clarify both these situations in more details?

A bug is actually stale if feedback or discussion is needed, which is not provided. The team has no way of finding out how to reproduce the bug, and cannot assess if it’s actually a bug or user-error. The same applies to usage-questions, which should not be raised as issues anyways. In these cases, closing after time-of-inactivity is legitimate in my opinion.

The other is a situation where not even triaging is possible due to external constraints. It would be great if these would not happen. If they do, though, I think there will be a higher chance that people resume work if only the most pressing news of that period remain to be immediately seen. It’s how human minds work: intractable heaps of work make people flee. We can agree this should not happen, and that we assume(!) this less likely to happen again. If it does, it’s pain for all involved parties. We have no way of knowing what’s the better way of handling unless we try it out.

> Currently the stale bot simply closes after some inactivity time. Inactivity does not indicate that the issue is "actually-stale bugs or questions that are never going to get resolved". It could indicated a few things. For example it could indicated that the dev team did not have time to triage or prioritise it. In this latter case, do you think it's a right thing to close the issue, and if yes, why?

It’s the latter case in your question that is the big point of criticism here. In our recent meetings, we agreed that these situations are horrible. Resuming project-work after long project-inactivity with tons and tons of lone triaging is horrible. People tend not to do it. That’s a downward-sprial making things even worse. If these situations happen, we believe it’s at least a viable experiment to auto-close things after some time elapsed in which noone triaged/prioritized it. That way, at least, issue-authors know that something has gone horribly wrong. And contributors/maintainers will have less of a hard time to pick up the keyboard again. Agreed — there is loss of work and loss of knowledge. But growing heaps of un-triaged issues also means a great productivity loss … so this is our attempt to try something out that may work.

> > Being a emergency stopper in the hopefull-never-again-phases where noone can even look at incoming issues.
> 
> How does this work? Could you please provide some more details?

If there is not even enough time to read and label issues, nothing will happen on that issue in the foreseeable future. It’s fair to let the user know. It’s fair to unburden the team a bit. We do not expect it to happen. If it happens, all possible actions have severe downsides. We chose one path of action that we believe could work better than the current model.


> The main difference is that if /lifecycle frozen worked I could trigger it, and the abovementioned labels you have to trigger. So I feel I'm left without any agency at all here. To re-iterate, I feel it is okay, to close an issue (after a discussion perhaps), the dev team believes won't fit. It is not okay to close it, because some arbitrary length of time has passed. Would you agree with that?

I think we are in agreement that closing things after arbitrary delays is a bad thing to do. That is why we are currently implementing this as a last resort option for "bad times". We believe that while it’s "not okay", there may be situations where it’s the only way out. We don’t want them to happen. We don’t assume they will happen. However, we see the risk, and prepare.

I still don’t fully understand that `/lifecycle frozen` thing, since noone ever pointed us to a piece of code which implements it, nor is there any description of what it does. I also don’t think that anybody should be able to randomly triage issues: Especially novice users often mis-use these powers — while experienced users and developers often are very careful with them. Since we get a good mix in our issues, I believe this is not something smart to consider today.



>     Is stale bot going to continue to close issue because some arbitrary length of time has passed?

Nitpicky answer: No, because stale has not closed any issues yet.
Proper answer: Yes, but only if we fail to label triage issues beforehand. This is not expected to happen, see above.

>     What are going to be the rules / setting for the stale bot?

We are still actively shaping our triaging and review process. As such, we are still adapting its config as part of this effort. Please check https://github.com/Mailu/Mailu/commits/master/.github/stale.yml for details, and also feel free to join either one of our project meetings or drop by in the matrix channel.

>     Will you allow your user base to /lifecycle frozen, when you yourselves do not have time to look / triage?

See the above elaborations on this. My personal opinion is against it, but if you provide some details, we are open to discuss the possibility in some of the upcoming meetings.

**Comment by briantopping** [comment]
This discussion really has turned around my perspective on what is being implemented. thank you. I hope that combined community member perspective helped evoke the generous effort to write this rather than alienate contributors. OSS obviously does not survive without contributors, there is no dominion over other contributors, and it has to be fun or nobody wants to work on it. Projects go in all kinds of directions. As long as frustration can be tolerated and excused, no apology is necessary for a lack of elaboration before this point.

In the end, I hope that content from here can be copy-pasted into a "contributors page" or simply included by reference. The opinion of this contributor is that the methodology is sound, including @AndrewSav's comments about agency. Community members who are invited to become project members **should** have agency around PM decisions or it just simply takes too much PM effort, which isn't much fun for development projects. And stalebot **is** appropriate for feature requests where context or clarifications are not available from the original author, for instance. In all, the vision for this is much clearer than "just add stalebot" and let things sort themselves out. 

In summary, I hope those concerned have similar perspectives and apologize for any stress that I may have added in bringing up these points.

**Comment by AndrewSav** [comment]
@Nebukadneza Thank you for this.

For me the principal disagreement with you, is that a mound of issues is somehow depressing to the degree, that the team members cannot deal with that internally (presumably we are all grown up), and have to push some misery on users (with the best intentions to avoid this if possible at all).

My experience is that any moderately sized project will have a number of issues in hundreds if not thousands. Many of those  will never be resolved. I cannot see how closing these issues helps the developers at all, I see how it frustrates users though. Issues that _for sure_ never will be worked on (unclear what is being asked, no answer given when question asked, features that does not fit philosophy of the project, etc) of course has to be closed. For the rest of them, who know, who and when decides to submit a helpful unexpected PR?

There is no point arguing though (unless you want to keep discussion going, of course, I'm happy to keep talking, it's just at this stage it does not look like it will change much), I think I understand your point of view, and mine is different. I'll be closing issues that I have opened, to help you keep things under control and less stressful, and I will be departing. I wish you and the team good luck.

**Comment by briantopping** [comment]
Hey @AndrewSav, I agree with what you are saying. Neither of us are on the team, and contributions of any form are expensive to personal time, which is what I think we are debating here. 

That said, I'm going to give the team a chance to fulfill the promises made above before walking away. I think the project is better than others I've found in the domain, but there's a lot of glaring problems as well (some that we've documented and are in contention through this defaulting process). And I don't have time to rewrite everything in Go with proper interfaces between modules and Kubernetes as a requirement, which probably wouldn't be too terribly difficult with Mailu as a template. 

Would it be something worth approaching as a team effort? Maybe. Would it pay the bills? I don't know about that. I get the feeling if it did, more people would magically have more time to work on it. There's no telling what can be done with the right team and product though. It just depends on taking an inventory of team goals, product requirements and risk. Even that is a chunk of work. 

I hope you might reconsider sticking around a bit or let me know where you are going if you are still going to jet. I appreciate the perspective and candor you brought into the conversation. It's healthy and important. Cheers!

**Comment by AndrewSav** [comment]
@briantopping I used to use https://github.com/hardware/mailserver before the original author could not spend time on it anymore. After that I evaluated a few solutions, and Mailu was the most likely candidate except for a few issues, that would be a regression for me.

Since that time I and a few other user of `hardware/mailserver` teamed up and and created a bare minimum maintenance fork: https://github.com/mailserver2/mailserver This is what I'm using right now.

Mailserver is solid, but it's a bit more complex to setup than mailu, and obviously, for any new features, someone has to work on them, which is, at this stage unlikely. For me Mailserver as an alternative to mailu makes sense, because I'm already pretty familiar with it, but I'm not so sure for other people, such as yourself. If you want to give it a try and have question, let me know by either opening an issue on the mailserver2 repo, or joining discord that I created for it, although it's pretty empty I check it regularly.

I have absolutely no hopes or aspirations that mailserver2 would pay the bills, it's just something that I used for many years, so it made sense to me to help with the bare minimum maintenance.

**Comment by Nebukadneza** [comment]
Hi again,

first off, sorry for the long silence.

I’m really glad to hear that @briantopping is not strongly opposed to our idea and experiment anymore. Thank you for being so collaborative and open-minded, I’m really happy about this.

At the same time, I’m sad to hear @AndrewSav s stance — but I don’t know what else could be said or done at the current point in time to resolve our conflict of ideas. In this light, I warmly wish you success in reviving mailserver2!

Since I also agree with @AndrewSav that this discussion has probably made all points that could have been made, I’ll close this issue. If you have further discussion, please nevertheless feel free to comment or reopen it. If there are any "new" misbehaviors or StaleBot, please also feel free to open a new issue.

Thanks everyone for their time participating in this discussion, and the project!
